shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/SafeHash.hs

```diff
@@ -0,0 +1,164 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE EmptyDataDecls #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+-- HeapWords instance
+{-# LANGUAGE UndecidableInstances #-}
+
+module Cardano.Ledger.SafeHash
+  ( SafeHash,
+    SafeToHash (originalBytes),
+    castSafeHash,
+    HashAnnotated,
+    hashAnnotated,
+    HashWithCrypto (..),
+    hashProxy,
+    HasAlgorithm,
+    extractHash,
+    EraIndependentTx,
+    EraIndependentTxBody,
+    EraIndependentBlockBody,
+    EraIndependentMetadata,
+    EraIndependentScript,
+    EraIndependentData,
+    EraIndependentScriptData,
+    EraIndependentAuxiliaryData,
+    unsafeMakeSafeHash,
+  )
+where
+
+import Cardano.Binary (FromCBOR (..), ToCBOR (..))
+import qualified Cardano.Crypto.Hash as Hash
+import qualified Cardano.Ledger.Crypto as CC
+import Cardano.Ledger.Era
+import Cardano.Prelude (HeapWords (..))
+import Control.DeepSeq (NFData)
+import Data.ByteString (ByteString)
+import Data.ByteString.Short (ShortByteString, fromShort)
+import Data.MemoBytes (MemoBytes (..))
+import Data.Typeable
+import NoThunks.Class (NoThunks (..))
+
+-- ==========================================================
+-- A SafeHash is a hash of something that is safe to hash.
+
+newtype SafeHash crypto index = SafeHash (Hash.Hash (CC.HASH crypto) index)
+  deriving (Show, Eq, Ord, SafeToHash, NoThunks, NFData)
+
+deriving newtype instance HeapWords (Hash.Hash (CC.HASH c) i) => HeapWords (SafeHash c i)
+
+deriving instance (Typeable index, CC.Crypto c) => ToCBOR (SafeHash c index)
+
+deriving instance (Typeable index, CC.Crypto c) => FromCBOR (SafeHash c index)
+
+-- We do NOT export the constructor SafeHash, but instead export
+-- 'hashAnnotated' and 'extractHash' which have constraints that limit
+-- their application to types which preserve their original serialization
+-- bytes. These are the functions people use.
+
+type HasAlgorithm c = Hash.HashAlgorithm (CC.HASH c)
+
+hashAnnotated :: forall c i x. (HasAlgorithm c, HashAnnotated x i c) => x -> SafeHash c i
+hashAnnotated x = hashProxy (Proxy @c) (indexProxy x) x
+
+extractHash :: SafeHash crypto i -> Hash.Hash (CC.HASH crypto) i
+extractHash (SafeHash h) = h
+
+-- Occasionally we need to change the HashAnnotated (which is a phantom type)
+
+castSafeHash :: forall i j c. SafeHash c i -> SafeHash c j
+castSafeHash (SafeHash h) = SafeHash (Hash.castHash h)
+
+-- Don't use this except in Testing to make Arbitrary instances, etc.
+unsafeMakeSafeHash :: (Hash.Hash (CC.HASH crypto) index) -> SafeHash crypto index
+unsafeMakeSafeHash x = SafeHash x
+
+-- =====================================================================
+-- Only Types that preserve their serialisation bytes are members of the
```

Comment:
A general point: details like this could be better expressed as haddock comments on the type (class) in question. That means that, for example, they show up in haddock if people search for this.

Reply:
Good idea.